File: apps/worker/src/agent_voice_worker/protocol/parse.py
"""Decode, validate, and encode messages that cross the data channel.

Mirrors `packages/protocol/src/parse.ts`. Never throws on malformed,
oversized, unknown, or unsupported input; every parse function returns a
typed result so callers can log and move on.
"""

from __future__ import annotations

import json
from collections.abc import Callable
from dataclasses import dataclass, field
from typing import Generic, Literal, TypeVar

from pydantic import BaseModel, ValidationError

from agent_voice_worker.protocol.commands import COMMAND_MODELS, _Command
from agent_voice_worker.protocol.envelope import Envelope
from agent_voice_worker.protocol.events import EVENT_MODELS, _Event
from agent_voice_worker.protocol.limits import LIMITS, PROTOCOL_VERSION

T = TypeVar("T")

MAX_ISSUES = 10
MAX_ISSUE_CHARS = 200
MAX_TYPE_ECHO_CHARS = 64

ParseReason = Literal[
    "invalid_json", "too_large", "unsupported_version", "unknown_event", "invalid_event"
]


@dataclass(frozen=True, slots=True)
class Ok(Generic[T]):
    value: T
    ok: Literal[True] = field(default=True, init=False)


@dataclass(frozen=True, slots=True)
class Err:
    reason: ParseReason
    ok: Literal[False] = field(default=False, init=False)
    bytes: int | None = None
    version: int | None = None
    type: str | None = None
    issues: tuple[str, ...] | None = None


ParseResult = Ok[T] | Err


def _utf8_byte_length(text: str) -> int:
    return len(text.encode("utf-8"))


def _format_issues(error: ValidationError) -> tuple[str, ...]:
    issues: list[str] = []
    for issue in error.errors()[:MAX_ISSUES]:
        path = ".".join(str(part) for part in issue["loc"])
        text = f"{path}: {issue['msg']}" if path else issue["msg"]
        if len(text) > MAX_ISSUE_CHARS:
            text = f"{text[: MAX_ISSUE_CHARS - 1]}…"
        issues.append(text)
    return tuple(issues)


def decode_raw(input_value: object) -> ParseResult[object]:
    """Turns a raw payload (JSON text, UTF-8 bytes, or an already-decoded
    value) into a plain value, enforcing the byte budget first."""
    if isinstance(input_value, bytes | bytearray | memoryview):
        raw = bytes(input_value)
        if len(raw) > LIMITS.max_event_bytes:
            return Err(reason="too_large", bytes=len(raw))
        try:
            input_value = raw.decode("utf-8")
        except UnicodeDecodeError:
            return Err(reason="invalid_json")
    if isinstance(input_value, str):
        byte_length = _utf8_byte_length(input_value)
        if byte_length > LIMITS.max_event_bytes:
            return Err(reason="too_large", bytes=byte_length)
        try:
            return Ok(json.loads(input_value))
        except json.JSONDecodeError:
            return Err(reason="invalid_json")
    return Ok(input_value)


def _check_envelope(value: object) -> ParseResult[dict[str, object]]:
    """Validates the version and the shared envelope fields, in that order."""
    if not isinstance(value, dict):
        return Err(reason="invalid_event", issues=("payload must be a JSON object",))
    version = value.get("v")
    if version != PROTOCOL_VERSION:
        return Err(
            reason="unsupported_version", version=version if isinstance(version, int) else None
        )
    try:
        Envelope.model_validate(value)
    except ValidationError as error:
        return Err(reason="invalid_event", issues=_format_issues(error))
    return Ok(value)


M = TypeVar("M", bound=BaseModel)


def _parse_message(
    input_value: object,
    models: dict[str, type[M]],
) -> ParseResult[M]:
    decoded = decode_raw(input_value)
    if not decoded.ok:
        return decoded
    checked = _check_envelope(decoded.value)
    if not checked.ok:
        return checked

    message_type = str(checked.value.get("type"))
    model = models.get(message_type)
    if model is None:
        return Err(reason="unknown_event", type=message_type[:MAX_TYPE_ECHO_CHARS])
    try:
        return Ok(model.model_validate(checked.value))
    except ValidationError as error:
        return Err(reason="invalid_event", issues=_format_issues(error))


def parse_event(input_value: object) -> ParseResult[_Event]:
    """Parses and validates an event received from the worker side."""
    return _parse_message(input_value, EVENT_MODELS)


def parse_command(input_value: object) -> ParseResult[_Command]:
    """Parses and validates a command received from the UI."""
    return _parse_message(input_value, COMMAND_MODELS)


class ProtocolEncodeError(Exception):
    def __init__(self, failure: Err, message: str) -> None:
        super().__init__(message)
        self.failure = failure


def _encode_validated(value: M, parse: Callable[[object], ParseResult[M]]) -> str:
    checked = parse(value.model_dump(mode="json", exclude_none=True))
    if not checked.ok:
        raise ProtocolEncodeError(checked, f"refusing to encode message: {checked.reason}")
    text = json.dumps(checked.value.model_dump(mode="json", exclude_none=True))
    byte_length = _utf8_byte_length(text)
    if byte_length > LIMITS.max_event_bytes:
        raise ProtocolEncodeError(
            Err(reason="too_large", bytes=byte_length),
            f"refusing to encode message: {byte_length} bytes exceeds {LIMITS.max_event_bytes}",
        )
    return text


def encode_event(event: _Event) -> str:
    """Validates and serialises an event. Raises `ProtocolEncodeError`
    rather than emit junk."""
    return _encode_validated(event, parse_event)


def encode_command(command: _Command) -> str:
    """Validates and serialises a command. Raises `ProtocolEncodeError`
    rather than emit junk."""
    return _encode_validated(command, parse_command)

File: packages/protocol/src/parse.ts
import type { z } from 'zod';

import { commandSchemas, isCommandType, type AgentVoiceCommand } from './commands.js';
import { envelopeSchema } from './envelope.js';
import { eventSchemas, isEventType, type AgentVoiceEvent } from './events.js';
import { LIMITS, PROTOCOL_VERSION } from './limits.js';

export type ParseFailure =
  | { ok: false; reason: 'invalid_json' }
  | { ok: false; reason: 'too_large'; bytes: number }
  | { ok: false; reason: 'unsupported_version'; version: number | null }
  | { ok: false; reason: 'unknown_event'; type: string }
  | { ok: false; reason: 'invalid_event'; issues: string[] };

export type ParseResult<T> = { ok: true; value: T } | ParseFailure;

/** JSON text, UTF-8 bytes, or an already-decoded value. */
export type RawMessage = unknown;

const MAX_ISSUES = 10;
const MAX_ISSUE_CHARS = 200;
const MAX_TYPE_ECHO_CHARS = 64;
const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder('utf-8', { fatal: true });

export function utf8ByteLength(text: string): number {
  return textEncoder.encode(text).byteLength;
}

/** Formats zod issues into short, bounded, path-prefixed strings. */
export function formatIssues(error: z.ZodError): string[] {
  return error.issues.slice(0, MAX_ISSUES).map((issue) => {
    const path = issue.path.map(String).join('.');
    const text = path ? `${path}: ${issue.message}` : issue.message;
    return text.length > MAX_ISSUE_CHARS ? `${text.slice(0, MAX_ISSUE_CHARS - 1)}…` : text;
  });
}

function isArrayBufferLike(value: unknown): value is ArrayBuffer {
  const tag = Object.prototype.toString.call(value);
  return tag === '[object ArrayBuffer]' || tag === '[object SharedArrayBuffer]';
}

/**
 * Turns a raw payload (JSON text, UTF-8 bytes, or an already-decoded value)
 * into a plain value, enforcing the byte budget first.
 */
export function decodeRaw(input: RawMessage): ParseResult<unknown> {
  // Brand checks rather than `instanceof`: bytes that crossed a realm
  // boundary (iframe, worker, test DOM) are still bytes.
  if (isArrayBufferLike(input)) {
    input = new Uint8Array(input);
  } else if (ArrayBuffer.isView(input)) {
    input = new Uint8Array(input.buffer, input.byteOffset, input.byteLength);
  }
  if (input instanceof Uint8Array) {
    if (input.byteLength > LIMITS.maxEventBytes) {
      return { ok: false, reason: 'too_large', bytes: input.byteLength };
    }
    try {
      input = textDecoder.decode(input);
    } catch {
      return { ok: false, reason: 'invalid_json' };
    }
  }
  if (typeof input === 'string') {
    const bytes = utf8ByteLength(input);
    if (bytes > LIMITS.maxEventBytes) {
      return { ok: false, reason: 'too_large', bytes };
    }
    try {
      return { ok: true, value: JSON.parse(input) as unknown };
    } catch {
      return { ok: false, reason: 'invalid_json' };
    }
  }
  return { ok: true, value: input };
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

/** Validates the version and the shared envelope fields, in that order. */
export function checkEnvelope(value: unknown): ParseResult<Record<string, unknown>> {
  if (!isRecord(value)) {
    return { ok: false, reason: 'invalid_event', issues: ['payload must be a JSON object'] };
  }
  if (value.v !== PROTOCOL_VERSION) {
    const version = typeof value.v === 'number' ? value.v : null;
    return { ok: false, reason: 'unsupported_version', version };
  }
  const envelope = envelopeSchema.safeParse(value);
  if (!envelope.success) {
    return { ok: false, reason: 'invalid_event', issues: formatIssues(envelope.error) };
  }
  return { ok: true, value };
}

function parseMessage<T>(
  input: RawMessage,
  isKnown: (type: string) => boolean,
  schemaFor: (type: string) => z.ZodType<T>,
): ParseResult<T> {
  const decoded = decodeRaw(input);
  if (!decoded.ok) return decoded;
  const envelope = checkEnvelope(decoded.value);
  if (!envelope.ok) return envelope;

  const type = String(envelope.value.type);
  if (!isKnown(type)) {
    return { ok: false, reason: 'unknown_event', type: type.slice(0, MAX_TYPE_ECHO_CHARS) };
  }
  const result = schemaFor(type).safeParse(envelope.value);
  if (!result.success) {
    return { ok: false, reason: 'invalid_event', issues: formatIssues(result.error) };
  }
  return { ok: true, value: result.data };
}

/**
 * Parses and validates an event received from the worker. Never throws:
 * every malformed, oversized, unknown, or unsupported message comes back as a
 * typed failure so consumers can log and move on.
 */
export function parseEvent(input: RawMessage): ParseResult<AgentVoiceEvent> {
  return parseMessage<AgentVoiceEvent>(
    input,
    isEventType,
    (type) => eventSchemas[type as keyof typeof eventSchemas] as z.ZodType<AgentVoiceEvent>,
  );
}

/** Parses and validates a command received from the UI. Same guarantees as `parseEvent`. */
export function parseCommand(input: RawMessage): ParseResult<AgentVoiceCommand> {
  return parseMessage<AgentVoiceCommand>(
    input,
    isCommandType,
    (type) => commandSchemas[type as keyof typeof commandSchemas] as z.ZodType<AgentVoiceCommand>,
  );
}

export class ProtocolEncodeError extends Error {
  constructor(
    readonly failure: ParseFailure,
    message: string,
  ) {
    super(message);
    this.name = 'ProtocolEncodeError';
  }
}

function encodeValidated<T>(value: T, parse: (input: unknown) => ParseResult<T>): string {
  const checked = parse(value);
  if (!checked.ok) {
    throw new ProtocolEncodeError(checked, `refusing to encode message: ${checked.reason}`);
  }
  const text = JSON.stringify(checked.value);
  const bytes = utf8ByteLength(text);
  if (bytes > LIMITS.maxEventBytes) {
    throw new ProtocolEncodeError(
      { ok: false, reason: 'too_large', bytes },
      `refusing to encode message: ${bytes} bytes exceeds ${LIMITS.maxEventBytes}`,
    );
  }
  return text;
}

/** Validates and serialises an event. Throws `ProtocolEncodeError` rather than emit junk. */
export function encodeEvent(event: AgentVoiceEvent): string {
  return encodeValidated(event, parseEvent);
}

/** Validates and serialises a command. Throws `ProtocolEncodeError` rather than emit junk. */
export function encodeCommand(command: AgentVoiceCommand): string {
  return encodeValidated(command, parseCommand);
}
